pkg/exchange/bitget/stream.go

```diff
@@ -0,0 +1,167 @@
+		return fmt.Errorf("unexpected subscription type: %v", opType)
+	}
+
+	logger := log.WithField("opType", opType)
+	args := []WebsocketArg{}
+	for _, subscription := range s.Subscriptions {
+		arg, err := s.convertSubscription(subscription)
+		if err != nil {
+			logger.WithError(err).Errorf("convert error, subscription: %+v", subscription)
+			return err
+		}
+
+		args = append(args, arg)
+	}
+
+	logger.Infof("%s channels: %+v", opType, args)
+	if err := s.Conn.WriteJSON(WsOp{
+		Op:   opType,
+		Args: args,
+	}); err != nil {
+		logger.WithError(err).Error("failed to send request")
+		return err
+	}
+
+	return nil
+}
+
+func (s *Stream) Unsubscribe() {
+	// errors are handled in the syncSubscriptions, so they are skipped here.
+	_ = s.syncSubscriptions(WsEventUnsubscribe)
+	s.Resubscribe(func(old []types.Subscription) (new []types.Subscription, err error) {
+		// clear the subscriptions
+		return []types.Subscription{}, nil
+	})
+}
+
+func (s *Stream) createEndpoint(_ context.Context) (string, error) {
+	var url string
+	if s.PublicOnly {
+		url = bitgetapi.PublicWebSocketURL
+	} else {
+		url = bitgetapi.PrivateWebSocketURL
+	}
+	return url, nil
+}
+
+func (s *Stream) dispatchEvent(event interface{}) {
+	switch e := event.(type) {
+	case *WsOpEvent:
+		if err := e.IsValid(); err != nil {
+			log.Errorf("invalid event: %v", err)
+		}
+
+	case *BookEvent:
+		s.EmitBooksEvent(*e)
+	}
+}
+
+func (s *Stream) parseWebSocketEvent(in []byte) (interface{}, error) {
+	var e WsEvent
+
+	err := json.Unmarshal(in, &e)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case e.IsOp():
+		return e.WsOpEvent, nil
+
+	case e.IsChannel():
+		switch e.Arg.Channel {
+		case ChannelOrderBook, ChannelOrderBook5, ChannelOrderBook15:
+			var books BookEvent
+			err = json.Unmarshal(e.WsChannelEvent.Data, &books.Events)
+			if err != nil {
+				return nil, fmt.Errorf("failed to unmarshal data into BookEvent: %+v, err: %w", string(e.WsChannelEvent.Data), err)
+			}
+
+			books.Type = e.WsChannelEvent.Action
+			books.instId = e.Arg.InstId
+			return &books, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unhandled websocket event: %+v", string(in))
+}
+
+func (s *Stream) handlerConnect() {
+	if s.PublicOnly {
+		// errors are handled in the syncSubscriptions, so they are skipped here.
+		_ = s.syncSubscriptions(WsEventSubscribe)
+	} else {
+		log.Error("*** PRIVATE API NOT IMPLEMENTED ***")
+	}
+}
+
+func (s *Stream) convertSubscription(sub types.Subscription) (WebsocketArg, error) {
+	w := WebsocketArg{
+		// support spot only
+		InstType: instSp,
+		Channel:  "",
+		InstId:   sub.Symbol,
+	}
+
+	switch sub.Channel {
+	case types.BookChannel:
+		w.Channel = ChannelOrderBook5
+
+		switch sub.Options.Depth {
+		case types.DepthLevel15:
+			w.Channel = ChannelOrderBook15
+		case types.DepthLevel200:
+			log.Info("*** The subscription events for the order book may return fewer than 200 bids/asks at a depth of 200. ***")
```

Comment:
It's better to add the exchange name "bitget" in this log.

Reply:
The log already with the filed 

https://github.com/c9s/bbgo/blob/5f8bda7d72933772b200bd746e5923361ec0e5eb/pkg/exchange/bitget/exchange.go#L14